### DISCUSSION TITLE
Setting content visibility according to audience

### DISCUSSION THREAD

**Discussion by monica-m-ps** [facebook/docusaurus#8050] [Ideas]
We have users like developers, partners, internal, external, customers. I need to set visibility of content according to my audience. We could have a login page and then a dashboard where we can limit some content etc. This'd help a lot.

**Comment by slorber** [comment]
Docusaurus builds a static site: the HTML files served to users are the same for users, they can not be different on a per-user basis.

At best you could create a Docusaurus site with empty pages, where the content only loads later (showing a spinner) through an authenticated API.

You can also build your docusaurus site in multiple static variants, and configure your host to serve one or the other according to the auth static of the user. Edge functions and hosting providers can allow that.

If you really need very dynamic content everywhere, maybe you shouldn't use Docusaurus in the first place and adopt something else with proper server-side-rendering support like Next.js or Remix

---

See also related another auth-related answer: https://github.com/facebook/docusaurus/issues/958#issuecomment-1057075925

**Reply by ezbeazy** [reply]
@slorber love your input on my suggested workaround. Dumb or possible?

**Comment by ezbeazy** [comment]
I think I found a work around for this. 

You can swizzle the root, like you would to redirect to a 3rd party authentication service, then use `useLocation` from `@docusaurus/router` to get the document location and only authenticate for documents in specific directories. 

Info on swizzling root & 3rd party auth
https://docusaurus.io/docs/swizzling#wrapper-your-site-with-root
https://stackoverflow.com/questions/71395788/can-firebase-authentication-work-with-docusaurus
https://medium.com/@thomasdevshare/docusaurus-authentication-with-firebase-c824da24bc51

Docusaurus Routing
https://docusaurus.io/docs/advanced/routing#generating-and-accessing-routes

More reference discussions on authentication w/ Docusaurus
https://github.com/facebook/docusaurus/issues/2769
https://github.com/facebook/docusaurus/issues/3748

**Reply by slorber** [reply]
@ezbeazy yes you can swizzle the Root if you want and eventually not render `children` if user is not authenticated, but this remains client-side code, a bit similar to what I suggest here:

> At best you could create a Docusaurus site with empty pages, where the content only loads later (showing a spinner) through an authenticated API.

Not that with your approach, even if the content doesn't "appear" on the site unless user is authenticated, the content still exist in JS files (mdx is compiled to JS) so it remains accessible for a technical person: it may be a security problem you have to consider.

If you want to be sure to be secure, you have to make sure the content is fully unavailable (not just in the presented html, but also in the JS code): fetching the sensitive content from a remote source (CMS) on the client-side is likely a better option. It also gives you the ability to have hybrid docs where some content is public/static, and some blocks are loaded on demand. This is more flexible than having a doc being fully public or private.

**Reply by ezbeazy** [reply]
@slorber good point. Is that only true of Docusaurus or most SSG's? 

any suggestions for integrating Docusaurus into a CMS or framework like Next?

**Reply by slorber** [reply]
@ezbeazy it's true for most SSGs: if sensitive content can be discovered through inspecting JS, then it can leak. 

You can integrate Docusaurus with a CMS by downloading markdown files locally as a pre-build step, we'll make it easier.

You can't integrate Docusaurus with Next.js, unless you build it statically upfront and then ask Next.js to serve the Docusaurus-built static files This may change in the future if we succeed in making Docusaurus a portable solution that can run on another framework, maybe like a Next.js plugin but that's an ambitious goal.

**Reply by ezbeazy** [reply]
@slorber, thank you for the explanation. That's actually the 'integration' method I was referring to, a mono-repo-like setup. Seemless integration into another framework would, to me, change the Docusaurus product itself from a framework to a tool or utility. Although, spinning out some features of Docusaurus into framework agnostic packages might be cool. 

One other thing. I was self conscious of asking yesterday, and forgive my ignorance and tangent, but to go a little deeper on the Docusaurus architecture... I'm trying to understand why the document data exists in the JS and is accessible client-side (say in the Sources tab of Dev Tools). 

My initial understanding was that JS and MDX files were compiled/built on the server into static HTML files to be served to the user within the root component. Not So?  

After reading https://docusaurus.io/docs/advanced/ssg; Does Docusaurus consider the Docs files/data "dynamic" parts of the site? Basically, client-side correlates the virtual DOM with the static DOM of the static HTML markup and then client-side JS containing mdx data is hydrated into the static HTML DOM? So is Docusaurus really only building and serving static markup skeleton and not a full-content containing static html page?

**Reply by slorber** [reply]
Docusaurus pre-builds html files for the initial page load only. If you navigate to /page2, it doesn't fetch the page2.html file but instead load React code for the client app to display page 2 content with JavaScript only.

The MDX compiler compiles Markdown to React components. We pre-render those, but it remains a React app.

**Reply by ezbeazy** [reply]
thanks @slorber, also found your explanation here: https://github.com/facebook/docusaurus/issues/7213

I feel like Docusaurus 2.0 transition to a SPA kind of branches away from what I think most people consider to be a SSG, thats fine, but definitely seems like there is some community confusion between terms and architecture.

**Reply by slorber** [reply]
🤷‍♂️ SSG can create SPAs for a few years already, and Docusaurus is not innovating here. There's also Next.js or Gatsby which are more popular tools also supporting SSG SPA output.
